### DISCUSSION TITLE
radius_graph with CUDA

### DISCUSSION THREAD

**Discussion by PabloVD** [pyg-team/pytorch_geometric#4213] [Q&A] [answered]
Hi, I'm trying to understand what radius_graph is actually doing to obtain the list of edges.
I see that, when the cpu is used, it is just building a KDTree using the scipy implementation. However, from the documentation and the torch_cluster repository, it is not obvious to me which algorithm is chosen when CUDA is employed.
Which is the expected time complexity in that case, how does it scale with the number of nodes considered?
Many thanks in advance!

**Comment by rusty1s** [comment, marked as answer]
The CUDA implementation will not build a KDTree. It will leverage the "power of parallel processors" to check if every element is within `radius` of the central node. As such, it's non-parallel runtime complexity will be O(N^2). Given N parallel processors, it runtime is reduced to O(N).

**Edit:** [Here](https://github.com/rusty1s/pytorch_cluster/blob/master/csrc/cuda/radius_cuda.cu) is the link to the code which sits in `torch-cluster`.

**Reply by PabloVD** [reply]
Many thanks for your reply! Do you know whether there are more efficient ways to create the graph? Maybe e.g. the [radius search from Open3D-ML](http://www.open3d.org/docs/latest/python_api_in/open3d.ml.torch.layers.FixedRadiusSearch.html)?

**Reply by rusty1s** [reply]
I don't have any experience with the radius search function from Open3D, but this looks indeed advantageous in case you want to perform multiple queries on an already existing KDTree (instead of re-computing it every time from scratch).

**Reply by PabloVD** [reply]
Thanks again for your response.

Let me ask about another issue. To check the time complexity in torch_cluster, I performed several tests with `radius_graph`, generating a graph using cuda from similar distributions of 1k, 10k and 100k points.

From 1000 to 10k number of points, there is an increase of factor 10 in the time required to generate the graph, as expected from the linear complexity behavior. However, from 10k to 100k, the time required is increased by a factor ~100, instead of scaling linearly. Is this behavior expected?

**Reply by rusty1s** [reply]
Do you have a small script to reproduce this? We also made some big speed improvements to `radius` and `knn` in `torch-cluster`, and I am currently in the process of crafting a new release.

**Reply by PabloVD** [reply]
Sure, I find that behavior with the following script:

```
import torch
import time
from torch_cluster import radius_graph


# use GPUs if available
if torch.cuda.is_available():
    print("CUDA Available")
    device = torch.device('cuda')
else:
    print('CUDA Not Available')
    device = torch.device('cpu')

torch.manual_seed(12345)

num_points = 10000

points = torch.randn([num_points,3])
radius = 0.8
loop = False
max_num_neighbors = points.shape[0]-1

points.to(device)

time_start = time.time()

edge_index = radius_graph(points, r=radius, loop=loop, max_num_neighbors=max_num_neighbors)

time_infer = time.time()-time_start
print("Num points", num_points, "time",time_infer)
```

For different number of points I get the next results:
Num points 1000 time 0.033
Num points 10000 time 0.298
Num points 100000 time 43.408

Changing `max_num_neighbors` reduce the time, but the relative time scaling is the same.

I also found the same time scaling with other point clouds, not only these random distributions.

I am using torch 1.10.0+cu113 and torch_cluster 1.5.9

**Reply by rusty1s** [reply]
Thank you. In the meantime, I just released [`torch-cluster==1.6.0`](https://github.com/rusty1s/pytorch_cluster/releases/tag/1.6.0).

**Reply by PabloVD** [reply]
So great, thanks!

I just ran my script with that new version but I do not find a significant improvement of the time taken to form the graph (only a gain of <~ 1 ms for 1k and 10k particles). Should we expect a faster performance?

**Reply by rusty1s** [reply]
It looks like your script is somewhat wrong as you are benchmarking CPU performance since you never assigned `points.to(device)`. Also keep in mind to synchronize CUDA when benchmarking. Furthermore, setting `max_num_neighbors` to the size of the point cloud can heavely slow down your code and should be used with care. Here is the corrected script:
```python
import torch
import time
from torch_cluster import radius_graph


# use GPUs if available
if torch.cuda.is_available():
    print("CUDA Available")
    device = torch.device('cuda')
else:
    print('CUDA Not Available')
    device = torch.device('cpu')

torch.manual_seed(12345)

num_points = 10000

points = torch.randn([num_points,3])
radius = 0.8
loop = False
max_num_neighbors = points.shape[0]-1

points.to(device)

time_start = time.time()

edge_index = radius_graph(points, r=radius, loop=loop, max_num_neighbors=max_num_neighbors)

time_infer = time.time()-time_start
print("Num points", num_points, "time",time_infer)
```
This gives me:
```
  1000: 0.0012s
 10000: 0.013s
100000: 0.335s
```

**Reply by PabloVD** [reply]
It seems that points.to(device) is already in my snippet above, although I noticed that running `points.is_cuda` gives me `False`. I need to write `points=points.to(device)` to actually having it on cuda. With that, now I get faster results than before with 10000 points, although slower than yours:
Num points 10000 time 0.045
and the same time for 1000. I don't know why this is not so fast, may it has to do with the cuda/torch version?

By the way, using torch.cuda.Event(enable_timing=True) and torch.cuda.synchronize() I got the same results when computing the time.

Thank you again!
